core/src/main/java/com/graphhopper/routing/InstructionsOutgoingEdges.java

```diff
@@ -129,22 +128,22 @@ public boolean outgoingEdgesAreSlowerByFactor(double factor) {
         double tmpSpeed = getSpeed(currentEdge);
         double pathSpeed = getSpeed(prevEdge);
 
-        // Speed-Change on the path indicates, that we change road types, show instruction
-        if (pathSpeed != tmpSpeed || pathSpeed < 1) {
+        // speed change indicates that we change road types
+        if (Math.abs(pathSpeed - tmpSpeed) >= 1) {
             return false;
         }
 
         double maxSurroundingSpeed = -1;
 
-        for (EdgeIteratorState edge : visibleAlternativeTurns) {
+        for (EdgeIteratorState edge : allowedAlternativeTurns) {
             tmpSpeed = getSpeed(edge);
             if (tmpSpeed > maxSurroundingSpeed) {
                 maxSurroundingSpeed = tmpSpeed;
             }
         }
 
-        // Surrounding streets need to be slower by a factor
-        return maxSurroundingSpeed * factor < pathSpeed;
+        // surrounding streets need to be slower by a factor and call round() so that tiny differences are ignored
+        return Math.round(maxSurroundingSpeed * factor) < Math.round(pathSpeed);
```

Comment:
These changes are the most significant. The problem with the loop over visibleAlternativeTurns was that it partially reads the speeds from edges with access=false which is possible when reading from an EncodedValue but not when calling calcWeight has we have a special protection there (i.e. it throws an exception).

The follow up changes regarding the round() call were then necessary to avoid (more or less) instructions just because of a tiny speed difference (<0.1km/h).